### PR TITLE
feat(schemas) Introduce column modifiers and stop making such modifier extend ColumnType

### PR DESCRIPTION
**Pull request by fpacifici**
While trying to refactor the column classes to support schemas with untyped columns (the SELECT clause of a subquery is a ColumnSet with untyped columns) I figured out I could not separate the hierarchy of DDL column types from the ones we use during query processing With `ColumnTypeWithModifier` in the way.

So I transformed Column modifiers into objects referenced by a column type instead of being column types themselves. 
This way a modifier is an attribute of a column type (Nullable is an attribute of a String column instead of being a type itself.), which seems to me more in line with what modifiers represent.

Basically the interesting changes are in clickhouse/columns.py and migrations/columns.py. The rest is a gigantic codemod to adapt all schemas and migrations.

### CODE REVIEW

File: snuba/clickhouse/columns.py [hunk 1]
```diff
@@ -1,21 +1,82 @@
...
+        """
+        The modifiers list to apply to this type. Modifiers are applied
+        to the type in the same order they appear in the list.
+        `String([LowCardinality(), Nullable()])`
+        means
+        `LowCardinality(Nullable(String))`
+        """
+        self.__modifiers = modifiers or []
+
+    def __repr__(self) -> str:
```
Comment (evanh) [inline snuba/clickhouse/columns.py:48]: Should this and `__eq__` include the modifiers as well?

**Review by evanh — COMMENTED** [review]
Just one question above

File: snuba/clickhouse/columns.py (hunk 1 above)
Comment (lynnagara) [reply snuba/clickhouse/columns.py:48]: Yeah, it appears that we are no longer comparing the columns + modifiers like we used to

**Comment by lynnagara** [comment]
Before this change, ColumnTypeWithModifier was internal to DDL only, and not supposed to leak outside of the migration code. My idea was to prevent these from being used in storage schemas via the type system, and only use the simplified column types in storage schemas. Now we are going in the opposite direction - replacing it with TypeModifier and using that everywhere. I think it makes the storage schemas more complex than they need to be, and enables the user to accidentally use migration type modifiers in storage schemas, which they should not do.

I also think we should also try to do some validation on modifiers, such as:
- limit the modifiers allowed for each column type (e.g. no low cardinality enums)
- enforce the order of the modifiers (e.g. you can't put Nullable after WithDefault)
- ensure there are no duplicates, (e.g. multiple Materialize statements).

This was always an issue before this change, but in some ways the correct way to write these feels even less obvious now that the syntax is more removed from the underlying ClickHouse syntax.

File: snuba/clickhouse/columns.py (hunk 1 above)
Comment (fpacifici) [reply snuba/clickhouse/columns.py:48]: Mh, this is weird. If this was not comparing modifiers, a test around parsing the schema should fail. Let me check what is going on there.

**Comment by fpacifici** [comment]
> My idea was to prevent these from being used in storage schemas via the type system, and only use the simplified column types in storage schemas. Now we are going in the opposite direction - replacing it with TypeModifier and using that everywhere.

I should have mentioned this. Actually I am planning as the next step to entirely split the hierarchies. Not only modifiers, but also types that you can use in schema columns vs types you can use only in migration columns. I tried to that first but I could not because of how get_raw works, thus I had to introduce the modifier first instead. Just cannot fit the two changes in the same PR. Will add the separate hierarchies today stacked top of this one so we can look at the two together.

+ 1 on the validation.

**Comment by fpacifici** [comment]
Slightly different approach: The modifiers for a type are packed into a TypeModifiers structure (with some helper methods to keep them concise).
The results are that:
- it is impossible to provide an invalid series of modifiers for a type. Since the MigrationTypeModifiers/SchemaTypeModifiers classes define what gets applied first and it is structurally impossible to add a modifier twice.
- Should we need more flexibility (I doubt) we can always implement another subclass for TypeModifiers that captures the semantics of multiple modifiers applied together
- It will be fairly easy to make types generic with respect to the TypeModifiers. This will allow us to ensure we cannot use Migration type modifiers outside of migrations.

(also fixed the `__eq__` methods)

File: snuba/datasets/storages/events_common.py [hunk 2]
```diff
@@ -173,12 +171,12 @@
...
-        ("culprit", Nullable(String())),
+        ("culprit", String(nullable())),
         ("sdk_integrations", Array(String())),
         ("modules", Nested([("name", String()), ("version", String())])),
-        ("release", ReadOnly(Nullable(String()))),
-        ("dist", ReadOnly(Nullable(String()))),
-        ("user", ReadOnly(Nullable(String()))),
+        ("release", (String(Modifiers(nullable=True, readonly=True)))),
+        ("dist", (String(Modifiers(nullable=True, readonly=True)))),
+        ("user", (String(Modifiers(nullable=True, readonly=True)))),
```
Comment (lynnagara) [inline snuba/datasets/storages/events_common.py:179]: Unnecessary parens slipped in here

File: snuba/migrations/snuba_migrations/events/0003_errors.py [hunk 3]
```diff
@@ -1,100 +1,99 @@
...
-    Column("sdk_version", LowCardinality(Nullable(String()))),
+    Column("platform", String(lowcardinality())),
+    Column("environment", String(Modifiers(nullable=True, low_cardinality=True))),
+    Column("release", String(Modifiers(nullable=True, low_cardinality=True))),
+    Column("dist", String(Modifiers(nullable=True, low_cardinality=True))),
+    Column("ip_address_v4", IPv4(nullable())),
+    Column("ip_address_v6", IPv6(nullable())),
+    Column("user", (String(Modifiers(default="''")))),
+    Column("user_hash", UInt(64, Modifiers(materialized="cityHash64(user)"))),
+    Column("user_id", String(nullable())),
```
Comment (lynnagara) [inline snuba/migrations/snuba_migrations/events/0003_errors.py:40]: I am not a huge fan of having many different conventions for specifying modifiers here. I see that `nullable` is provided to keep things more concise, however I'd personally prefer something a bit more consistent.

If the goal is to make things really concise and make it as easy as possible to for anyone to write migrations, what about going with something like:
`String(nullable=True, low_cardinality=True)` 

where `String` is a migrations specific class (possibly a subclass of the schema's String)

Comment (fpacifici) [reply snuba/migrations/snuba_migrations/events/0003_errors.py:40]: I think the `String(nullable=True, low_cardinality=True)` is viable, though wouldn't it still be yet another convention? `String(Modifiers(nullable=True, low_cardinality=True))` would still be accepted.

In that sense I think I would prefer removing `nullable()` and `lowcardinality()` entirely and replace them with 
`String(Modifiers(nullable=True, low_cardinality=True))`

What doo you think ?

Comment (lynnagara) [reply snuba/migrations/snuba_migrations/events/0003_errors.py:40]: Well, I mean it's a different convention but it would be would be the only convention we'd use everywhere. Not just for string, but we'd have the equivalent for every other column type and modifier combination, e.g. we'd also have `UInt(8, nullable=True, codec=["NONE"])` and friends. We've remove `nullable()` / `lowcardinality()`. We'd also never import MigrationModifiers in the migration. `MigrationModifiers` would instead get created by the ColumnType class. We'd also use a similar syntax for schemas and SchemaModifiers.

I agree that we can remove `nullable()` and `lowcardinality()` entirely whichever way we go, but I do think there's value in making the schemas as simple and easy to write as possible.

Comment (fpacifici) [reply snuba/migrations/snuba_migrations/events/0003_errors.py:40]: > we'd also have UInt(8, nullable=True, codec=["NONE"]) and friends

I think this would become a problem. (See the PR on top of this one: https://github.com/getsentry/snuba/pull/1443). Doing something like that would require us to duplicate all column types. The migration columns could not extend the schema columns (SchemaModifiers and MigrationModifiers are not one the subclass of the other intentionally). 

In order to reuse types across schemas and migrations The constructor either has to be the same or the migration one must be an extension of the schema one. That way I think what we loose in higher verbosity is better than fully duplicating the types.

One alternative is to let the classes the way they are with this PR and add helper methods to instantiate them instead of helper methods to provide modifiers:
```
class String:
   ... whatever it is now in this PR ...

# In clickhouse.columns
def string(nullable: bool, readonly: bool) -> String:
    return String(Modifiers(nullable=....))

# in migrations.py
def string(nullable: bool, lowcardinality: bool, default: string, etc.) -> String:
    return String(MigrationModifiers(.....))
```

Though I would do this addition in a separate PR since this is already mroe than 1000 lines long. Here I would just remove nullable() and readonly().

What about this ?

**Comment by fpacifici** [comment]
Helpers have been removed

**Review by lynnagara — APPROVED** [review]
